src/stream.rs

```diff
@@ -0,0 +1,250 @@
+// Copyright (c) 2022 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::unpack::unpack;
+use anyhow::{anyhow, Result};
+use sha2::Digest;
+use std::fs;
+use std::io::{Cursor, Read};
+use std::path::{Path, PathBuf};
+use tokio::io::{AsyncRead, AsyncReadExt};
+
+const CAPACITY: usize = 32768;
+const DIGEST_SHA256: &str = "sha256";
+const DIGEST_SHA512: &str = "sha512";
+
+const ERR_BAD_UNCOMPRESSED_DIGEST: &str = "unsupported uncompressed digest format";
+
+pub trait DigestHasher {
+    fn digest_update(&mut self, buf: &[u8]);
+    fn digest_finalize(self) -> String;
+}
+
+#[derive(Clone, Debug)]
+pub enum LayerDigestHasher {
+    Sha256(sha2::Sha256),
+    Sha512(sha2::Sha512),
+}
+
+impl DigestHasher for LayerDigestHasher {
+    fn digest_update(&mut self, buf: &[u8]) {
+        match self {
+            LayerDigestHasher::Sha256(hasher) => {
+                hasher.update(buf);
+            }
+            LayerDigestHasher::Sha512(hasher) => {
+                hasher.update(buf);
+            }
+        }
+    }
+
+    fn digest_finalize(self) -> String {
+        match self {
+            LayerDigestHasher::Sha256(hasher) => {
+                format!("{}:{:x}", DIGEST_SHA256, hasher.finalize())
+            }
+            LayerDigestHasher::Sha512(hasher) => {
+                format!("{}:{:x}", DIGEST_SHA512, hasher.finalize())
+            }
+        }
+    }
+}
+
+// Wrap a flume channel with [`Read`](std::io::Read) support.
+// This can bridge the [`AsyncRead`](tokio::io::AsyncRead) from
+// decrypt/decompress and impl Read for unpack.
+struct ChannelRead {
+    rx: flume::Receiver<Vec<u8>>,
+    current: Cursor<Vec<u8>>,
+}
+
+impl ChannelRead {
+    fn new(rx: flume::Receiver<Vec<u8>>) -> ChannelRead {
+        ChannelRead {
+            rx,
+            current: Cursor::new(vec![]),
+        }
+    }
+}
+
+impl Read for ChannelRead {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        // Receive new buffer when we handled previous data.
```

Comment:
Do you mean `Receive new buffer when we finish handling previous data`?

Reply:
Yes, thanks, fixed now!